crates/primitives/src/sol.rs

```diff
@@ -0,0 +1,440 @@
+    }
+
+    /// Converts to `Self` from `Self::SolType`.
+    fn from_sol_type(value: Self::SolType) -> Self;
+}
+
+/// Maps an arbitrary Rust/ink! type to a Solidity ABI type equivalent for Solidity
+/// ABI encoding.
+///
+/// # Note
+/// Implementing this trait entails:
+/// - Declaring the equivalent Solidity ABI type via the `SolType` associated type. See
+///   the [docs for sealed `SolTypeEncode` trait][SolTypeEncode] for a table of Rust/ink!
+///   primitive types mapped to their equivalent Solidity ABI type.
+/// - Implementing the `to_sol_type` method which defines how to convert (preferably via a
+///   borrow) from `&self` to `&Self::SolType` (i.e. the Solidity ABI representation).
+///
+/// # Example
+///
+/// ```
+/// use ink_primitives::SolEncode;
+///
+/// // Example arbitrary type.
+/// struct MyType {
+///     size: u8,
+///     status: bool,
+/// }
+///
+/// // `SolEncode` implementation/mapping.
+/// impl<'a> SolEncode<'a> for MyType {
+///     // NOTE: Prefer reference based representation for better performance.
+///     type SolType = (&'a u8, &'a bool);
+///
+///     fn to_sol_type(&'a self) -> Self::SolType {
+///         (&self.size, &self.status)
+///     }
+/// }
+/// ```
+pub trait SolEncode<'a> {
+    /// Equivalent Solidity ABI type representation.
+    type SolType: SolTypeEncode;
+
+    /// Name of equivalent Solidity ABI type.
+    const SOL_NAME: &'static str =
+        <<Self::SolType as SolTypeEncode>::AlloyType as AlloySolType>::SOL_NAME;
+
+    /// Solidity ABI encode the value.
+    fn encode(&'a self) -> Vec<u8> {
+        <Self::SolType as SolTypeEncode>::encode(&self.to_sol_type())
+    }
+
+    /// Converts from `Self` to `Self::SolType` via either a borrow (if possible), or
+    /// a possibly expensive conversion otherwise.
+    fn to_sol_type(&'a self) -> Self::SolType;
+}
+
+macro_rules! impl_primitive_decode {
+    ($($ty: ty),+ $(,)*) => {
+        $(
+            impl SolDecode for $ty {
+                type SolType = $ty;
+
+                fn from_sol_type(value: Self::SolType) -> Self {
+                    value
+                }
+            }
+        )*
+    };
+}
+
+macro_rules! impl_primitive_encode {
+    ($($ty: ty),+ $(,)*) => {
+        $(
+            impl<'a> SolEncode<'a> for $ty {
+                type SolType = &'a $ty;
+
+                fn to_sol_type(&'a self) -> Self::SolType {
+                    self
+                }
+            }
+        )*
+    };
+}
+
+macro_rules! impl_primitive {
+    ($($ty: ty),+ $(,)*) => {
+        $(
+            impl_primitive_decode!($ty);
+
+            impl_primitive_encode!($ty);
+        )*
+    };
+}
+
+impl_primitive! {
+    // bool
+    bool,
+    // signed integers
+    i8, i16, i32, i64, i128,
+    // unsigned integers
+    u8, u16, u32, u64, u128, U256,
+    // string
+    String,
+    // address
+    Address,
+}
+
+// Rust array <-> Solidity fixed-sized array (i.e. `T[N]`).
```

Comment:
```suggestion
// Rust array ⭤ Solidity fixed-sized array (i.e. `T[N]`).
```

Looks nicer in the rendered docs.